_posts/2017-06-09-how-we-found-a-memory-leak.markdown

```diff
@@ -0,0 +1,164 @@
+---
+layout: post
+title:  "Our journey discovering a Node.js memory leak"
+date:   2017-06-09 17:00:00 +0200
+categories: javascript node.js memoryleak bug
+author: Florian Störkle
+---
+
+# Tracking down a Node.js memory leak
+
+In this blog post we want to share our experience of tracking down and fixing a memory leak in one of our Node.js applications. For several reasons, this proved quite difficult, and took us several days to fix.
+
+## Preface
+
+Our team built a Node.js application that is responsible for synchronizing logs from an external service provider to our internal log pooling system (Google Stackdriver Logs). Let's call this application `sync-logs-to-stackdriver`.
+
+To make it simple, the job periodically checks if there are new logs available in the external system. If that’s the case it syncs them to Stackdriver. If not, it waits a short amount of time and then checks again for new logs.
+
+In our internal container infrastructure, we have two instances of this service running:
+
+- a production instance (syncing ~ 100 logs/minute)
+- a development instance (syncing ~ 100 logs/day)
+
+
+## Discovery
+
+One day, we were noticed by our infrastructure team that one of applications, `sync-logs-to-stackdriver`,  killed and restarted automatically several times a day. They said the likely cause would be a memory leak.
+
+As we are using NewRelic for monitoring our Node.js applications, we quickly checked the health of the application there. And indeed, we could confirm the suspicion: we really had a memory leak. This memory usage graph of the development instance of the application illustrates that:
+
+![NewRelic memory monitoring](https://github.hc.ag/pages/checkin/img/2017-05-10-memory-leak/dev-memory-leak-crashes.png)
+
+As one can see, the memory usage increases without upper border, until the container is killed by our infrastructure. Then the container is respawned, and it starts all over again.
+
+Interestingly, the leak was present in our production instance, but had far smaller impact – it only ran out of memory every other week.
+
+We noticed one thing that were quite suprised of that point in time: The memory leak was in the so-called "non-heap" memory of Node – that is the memory that is used by Node itself and binary extensions, but not by application code.
```

Comment:
Hm unfortunately, I can't find any good resource on that.
Only [this article from NewRelic](https://blog.newrelic.com/2017/03/30/nodejs-vm-metrics-view-apm/) explains it the "New metrics, too" section:

> Non-Heap Memory used by Node for things like buffers and streams

Do you think that link would help nevertheless?